olive/passes/onnx/common.py

```diff
@@ -0,0 +1,142 @@
+    return {
+        "save_as_external_data": PassConfigParam(
+            type_=bool,
+            default_value=False,
+            description=(
+                "Serializes tensor data to separate files instead of directly in the ONNX file. Large models (>2GB)"
+                " may be forced to save external data regardless of the value of this parameter."
+            ),
+        ),
+        "all_tensors_to_one_file": PassConfigParam(
+            type_=bool,
+            default_value=True,
+            description=(
+                "Effective only if save_as_external_data is True. If true, save all tensors to one external file"
+                " specified by 'external_data_name'. If false, save each tensor to a file named with the tensor name."
+            ),
+        ),
+        "external_data_name": PassConfigParam(
+            type_=str,
+            default_value=None,
+            description=(
+                "Effective only if all_tensors_to_one_file is True and save_as_external_data is True. If not specified,"
+                " the external data file will be named with <model_path_name>.data"
+            ),
+        ),
+    }
+
+
+def model_proto_to_file(
+    model: onnx.ModelProto,
+    output_path: Union[str, Path],
+    save_as_external_data: Optional[bool] = False,
+    all_tensors_to_one_file: Optional[bool] = True,
+    external_data_name: Optional[Union[str, Path]] = None,
+) -> bool:
+    """
+    Save the ONNX model to the specified path.
+
+    :param model: The ONNX model to save.
+    :param output_path: The path to save the ONNX model to.
+    :param save_as_external_data: If True, save tensor data to separate files instead of directly in the ONNX file.
+        Large models (>2GB) may be forced to save external data regardless of the value of this parameter.
+    :param all_tensors_to_one_file: Effective only if save_as_external_data is True. If True, save all tensors to one
+        external file specified by 'external_data_name'. If False, save each tensor to a file named with the tensor
+        name.
+    :param external_data_name: Effective only if all_tensors_to_one_file is True and save_as_external_data is True.
+        If not specified, the external data file will be named with <model_path_name>.data
+
+    :return: True if the model has external data, False otherwise.
+    """
+    output_path = Path(output_path)
+    if output_path.exists():
+        logger.info(f"Deleting existing onnx file: {output_path}")
+        output_path.unlink()
+
+    # parent directory of .onnx file
+    output_dir = output_path.parent
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    if not save_as_external_data:
+        try:
+            # save model
+            onnx.save_model(model, str(output_path))
+        except ValueError as e:
+            # model is too large to save as a single file
+            if "Please use save_as_external_data to save tensors separately from the model file." in str(e):
```

Comment:
With a sufficiently large model I actually get an error like the following, which won't be caught by your `if` check.

![Screenshot 2023-04-14 151005](https://user-images.githubusercontent.com/3835701/232163544-cc2adaa1-8d97-427c-9bdf-4cba815b67e4.png)

Reply:
Oh, thanks for the heads up. I thought I would catch the error correctly because of this https://github.com/onnx/onnx/blob/main/onnx/__init__.py#L110 

I will test the model save logic with a large model. Do you have one you recommend or I could just create a large dummy model?

Reply:
You can probably run the SD example. The u-net is the big one:

```
python -m olive.workflows.run --config .\config_unet.json
```

I think this will run correctly even with onnxruntime 1.14. It may fail to evaluate at the end, but it should at least convert successfully.

Reply:
Thanks! I will try this model. I have to see why onnx.save_model doesn't catch the error properly. Looks like it is raising the error from protobuf instead of raising the linked error.

Reply:
Found this issue. `onnx` only added the check in version 1.13.0.

Reply:
@jstoecker I have updated the check and validated it against onnx 1.12.0 also.